core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -857,9 +862,11 @@ private[spark] class BlockManager(
 
     val startTimeMs = System.currentTimeMillis
     var blockWasSuccessfullyStored: Boolean = false
+    var exceptionWasThrown: Boolean = true
     val result: Option[T] = try {
       val res = putBody(putBlockInfo)
       blockWasSuccessfullyStored = res.isEmpty
+      exceptionWasThrown = false
       res
     } finally {
       if (blockWasSuccessfullyStored) {
```

Comment:
That said, I think we could simplify this by moving the non-error-case code into the `try` block. Let me do that now.